### ISSUE TITLE
Allow Feature Policy-based permission models

### ISSUE THREAD

**Issue by jan-ivar** [w3c/permissions#185]
As I understand it, one driver of Feature Policy is that most users don't understand iframes, e.g. today:
![crossdomainprefeaturepolicyhalf](https://user-images.githubusercontent.com/3136226/47121904-94beb200-d242-11e8-9df1-a78f7b304452.png)

Feature Policy in theory lets browsers implement simpler permission models for the above, like this:
![crossdomainpostfeaturepolicyhalf](https://user-images.githubusercontent.com/3136226/47121913-9be5c000-d242-11e8-9368-6cfc2c880e0d.png)

...where grant is to the top-level domain *nytimes.com*, which delegates w/ `<iframe allow=”camera”>`.

Importantly, users would no longer expect permission to be persisted for *nytimes.techsupport.fr*, but for *nytimes.com*.

Currently, [request permission to use](https://w3c.github.io/permissions/#request-permission-to-use) says: *"The user’s interaction may provide new information about the user’s intent for this realm and other realms with the same origin."*

This seems too limiting, since the *origin* of the *realm* is still *nytimes.techsupport.fr* here AFAIK.

Any ideas on how this may be written to allow for these new permission models?

**Comment by raymeskhoury** [comment]
> Currently, request permission to use says: "The user’s interaction may provide new information about the user’s intent for this realm and other realms with the same origin."
> 
> This seems too limiting, since the origin of the realm is still nytimes.techsupport.fr here AFAIK.

Even though the response to the permission request only provides new information for that realm, the user agent is free to decide how it returns a value when ["Reading the current Permission State"](https://www.w3.org/TR/permissions/#reading-current-states) (See step 4 in particular).

That means that the user agent, when query permission for the iframe, is free to say something like "nytimes.techsupport.fr doesn't have access but the top level frame (nytimes.com) does have access so grant access to the iframe"

We intentionally left this freedom in "Reading the current Permission State" so that user agents would have freedom for things like this, but also for things like implementing Crowd Deny.

**Comment by jan-ivar** [comment]
Ok, sounds like we're good then. Closing.

**Comment by jan-ivar** [comment]
> "nytimes.techsupport.fr doesn't have access but the top level frame (nytimes.com) does have access so grant access to the iframe"

@raymeskhoury Doesn't the current language still prevent persisting permission to nytimes.com times in this case?

Doesn't it need to say something like: *"The user’s interaction may provide new information about the user’s intent for this realm and other realms with the same origin, or other realms with the same origin as the origin used to grant the permission"* ?

**Comment by raymeskhoury** [comment]
I'm not sure I follow - perhaps spell it out with an example?

**Comment by jan-ivar** [comment]
Firefox user clicks on nytimes.com's (hypothetical) *Live video support*, and sees this prompt:
```
Will you allow nytimes.com to use your camera and microphone?

☐ Remember this decision

Don't Allow                                         Allow
```
They check `☑ Remember this decision` and click `Allow`.

They're done with that and now go to `nytimes.com`'s *Live video chat with author Ian Famous*, and they see the exact same prompt again, wondering why Firefox didn't remember their decision?

It's because the second request came instead from iframe `authors.fameo.us`, and even though Firefox was *"free to say something like "nytimes.techsupport.fr" doesn't have access but the top level frame (nytimes.com) does have access so grant access to the iframe"*, it's not allowed to **persist** said access to nytimes.com because it doesn't have *"new information about the user’s intent for"* nytimes.com, since that site is not *"this realm and other realms with the same origin"* (as nytimes.techsupport.fr).

**Comment by raymeskhoury** [comment]
Thanks, makes sense. What I was referring to previously was that you can really return anything you want when "Reading the current Permission State" which can account for this at the point the user visits nytimes.com again.

If we want to loosen what is allowed at the point the user accepts the prompt though, I'm ok with that too.

**Comment by jan-ivar** [comment]
> you can really return anything you want when "Reading the current Permission State"

Not really. There's [this sentence](https://w3c.github.io/permissions/#permission-state) in there: *"If there was a previous invocation of this algorithm with the same descriptor and settings, returning previousResult, and the UA has not received [new information about the user’s intent](https://w3c.github.io/permissions/#new-information-about-the-users-intent) since that invocation, return previousResult."*

The limited example I gave may not show it, so let me elaborate on it:

1. Say the user blocks cam+mic access in NYTimes's "tech support" section.
2. Then they go to the "live chat with author" section and grant persistent cam+mic access.
3. Then they return to the tech-support section. The spec sentence above blocks access.

To the end-user, this is all nytimes.com, so access should not have been blocked in 3. The user shouldn't even see a prompt in this case, since they effectively approved persistent access in 2.

**Comment by jyasskin** [comment]
Maybe try something like

> The user’s interaction may provide new information about the user’s intent for this realm, the realms of the [ancestor browsing contexts](https://html.spec.whatwg.org/multipage/browsers.html#ancestor-browsing-context) of this realm's Document, and other realms with the same origins as those.

I haven't read through the whole thing in detail again, so please double-check that that actually says the right thing and fixes the problem.

**Comment by marcoscaceres** [comment]
**Summary of Issue #185** (Thanks ChatGPT)

**Context:**  
Issue #185 focuses on whether the Permissions specification should explicitly address "transient" or "ephemeral" permissions—those granted temporarily rather than stored persistently.

**Key Points:**

1. **Ephemeral Permissions Concept:**  
   The discussion considers if the spec should acknowledge or define states in which permissions are granted only for the current session, or until certain conditions are met, rather than permanently.

2. **User Agent Variation:**  
   Different browsers may implement ephemeral permissions differently (e.g., permissions valid only for the current tab/session). The question is whether the spec should unify these behaviors or simply acknowledge their existence.

3. **Spec Changes and Requirements:**  
   There’s no current push for strict, normative changes. The inclination is to note that ephemeral permissions may exist, providing general guidance without imposing exact requirements.

4. **Impact on CR Transition:**  
   This issue does not seem to block moving the Permissions specification forward. If addressed, it would likely involve adding non-normative clarifications rather than mandated changes.

**Conclusion:**  
No major normative changes are required for Issue #185. The concept of ephemeral permissions might be acknowledged in the specification as a non-normative note, without blocking the transition to Candidate Recommendation.

**Comment by jyasskin** [comment]
@marcoscaceres Please stop pasting ChatGPT sludge into public issues. If it helps you summarize the discussion for yourself, fine, but anything you post here needs to have your personal endorsement that it's worth other humans' time to read.

**Comment by jan-ivar** [comment]
This seems overtaken by events: "realm" isn't mentioned anymore.

Unlike 6 years ago, keying permissions off the top-level origin seems the norm now, even captured in the [default permission key generation algorithm](https://w3c.github.io/permissions/#dfn-default-permission-key-generation-algorithm): _"1. Return settings's [top-level origin](https://html.spec.whatwg.org/multipage/webappapis.html#concept-environment-top-level-origin)."_

The only remaining nit might be [request permission to use](https://w3c.github.io/permissions/#dfn-request-permission-to-use) seems to require a bit of reading ahead to discover this: _"_

4. _If the user gives [express permission](https://w3c.github.io/permissions/#dfn-express-permission) to use the powerful feature, set current state to "[granted](https://w3c.github.io/permissions/#dom-permissionstate-granted)"; otherwise to "[denied](https://w3c.github.io/permissions/#dom-permissionstate-denied)". The user's interaction may provide [new information about the user's intent](https://w3c.github.io/permissions/#dfn-new-information-about-the-user-s-intent) for the [origin](https://html.spec.whatwg.org/multipage/browsers.html#concept-origin)._

5. _Let key be the result of [generating a permission key](https://w3c.github.io/permissions/#dfn-permission-key-generation-algorithm) with the [current settings object](https://html.spec.whatwg.org/multipage/webappapis.html#current-settings-object)."_

Step 4 talks about UX centered around an "origin", but we have to dive into step 5 to learn the invariant we need "new information" about might be the top-level origin (by default), which seems relevant to UX.

That said, I can't think of a better way to write this, so maybe this is clear enough?
